From de0b973fc7f8a8411271c20e7b70ceaa8757a029 Mon Sep 17 00:00:00 2001
From: Laisheng Hu <laisheng@codeaurora.org>
Date: Wed, 3 Jul 2019 17:07:47 +0800
Subject: [PATCH 33/42] gstomx enc: support downscale encoding

As old vpu has no such function, use macro to control downscale code

Change-Id: Ib234871bc019bbe6759af06871b802d0c7a650e1
Signed-off-by: Laisheng Hu <laisheng@codeaurora.org>
---
 omx/gstomxvideoenc.c | 71 +++++++++++++++++++++++++++++++++++++++++++++++++++-
 omx/gstomxvideoenc.h |  4 +++
 2 files changed, 74 insertions(+), 1 deletion(-)

diff --git a/omx/gstomxvideoenc.c b/omx/gstomxvideoenc.c
index 1f52524..87ddd2c 100644
--- a/omx/gstomxvideoenc.c
+++ b/omx/gstomxvideoenc.c
@@ -319,6 +319,10 @@ enum
   PROP_MIRROR,
   PROP_INTRA_REFRESH_MODE,
   PROP_INTRA_REFRESH_MBS,
+#ifdef _USE_TARGET_VPU554_
+  PROP_DOWNSCALE_WIDTH,
+  PROP_DOWNSCALE_HEIGHT
+#endif
 };
 
 /* FIXME: Better defaults */
@@ -356,7 +360,10 @@ enum
 #define GST_OMX_VIDEO_ENC_ROTATION_DEFAULT (0)
 #define GST_OMX_VIDEO_ENC_MIRROR_DEFAULT (0)
 #define GST_OMX_VIDEO_ENC_INTRA_REFRESH_MODE_DEFAULT (0x7fffffff)
-
+#ifdef _USE_TARGET_VPU554_
+#define GST_OMX_VIDEO_ENC_DOWNSCALE_WIDTH_DEFAULT (0xffffffff)
+#define GST_OMX_VIDEO_ENC_DOWNSCALE_HEIGHT_DEFAULT (0xffffffff)
+#endif
 
 /* class initialization */
 #define do_init \
@@ -634,6 +641,21 @@ gst_omx_video_enc_class_init (GstOMXVideoEncClass * klass)
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
           GST_PARAM_MUTABLE_READY));
 
+#ifdef _USE_TARGET_VPU554_
+  g_object_class_install_property (gobject_class, PROP_DOWNSCALE_WIDTH,
+      g_param_spec_uint ("downscale-width", "downscale width",
+          "downscale_width (0xffffffff=component default)",
+          0, G_MAXUINT, GST_OMX_VIDEO_ENC_DOWNSCALE_WIDTH_DEFAULT,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));
+  g_object_class_install_property (gobject_class, PROP_DOWNSCALE_HEIGHT,
+      g_param_spec_uint ("downscale-height", "downscale height",
+          "downscale_height (0xffffffff=component default)",
+          0, G_MAXUINT, GST_OMX_VIDEO_ENC_DOWNSCALE_HEIGHT_DEFAULT,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));
+#endif
+
   element_class->change_state =
       GST_DEBUG_FUNCPTR (gst_omx_video_enc_change_state);
 
@@ -703,6 +725,10 @@ gst_omx_video_enc_init (GstOMXVideoEnc * self)
   self->mirror = GST_OMX_VIDEO_ENC_MIRROR_DEFAULT;
   self->intra_refresh_mode = GST_OMX_VIDEO_ENC_INTRA_REFRESH_MODE_DEFAULT;
   self->intra_refresh_mbs = 0;
+#ifdef _USE_TARGET_VPU554_
+  self->downscale_width = GST_OMX_VIDEO_ENC_DOWNSCALE_WIDTH_DEFAULT;
+  self->downscale_height = GST_OMX_VIDEO_ENC_DOWNSCALE_HEIGHT_DEFAULT;
+#endif
 
   g_mutex_init (&self->drain_lock);
   g_cond_init (&self->drain_cond);
@@ -1240,6 +1266,33 @@ gst_omx_video_enc_open (GstVideoEncoder * encoder)
       }
     }
 
+#ifdef _USE_TARGET_VPU554_
+    if (self->downscale_width != 0xffffffff && self->downscale_height != 0xffffffff) {
+      QOMX_INDEXDOWNSCALAR downscalar_params;
+      GST_OMX_INIT_STRUCT(&downscalar_params);
+      if (self->enc) {
+        downscalar_params.nPortIndex = self->enc_out_port->index;
+        err = gst_omx_component_get_parameter(self->enc, (OMX_INDEXTYPE)OMX_QcomIndexParamVideoDownScalar,(OMX_PTR)&downscalar_params);
+        if (err != OMX_ErrorNone) {
+          GST_ERROR_OBJECT (self,
+              "Failed to get downscaleparameter: %s (0x%08x)",
+              gst_omx_error_to_string (err), err);
+        }
+        downscalar_params.bEnable = OMX_TRUE;
+        downscalar_params.nOutputWidth = self->downscale_width;
+        downscalar_params.nOutputHeight = self->downscale_height;
+        err = gst_omx_component_set_parameter(self->enc, (OMX_INDEXTYPE)OMX_QcomIndexParamVideoDownScalar,(OMX_PTR)&downscalar_params);
+        if (err != OMX_ErrorNone) {
+          GST_ERROR_OBJECT (self,
+              "Failed to set downscale parameter: %s (0x%08x)",
+              gst_omx_error_to_string (err), err);
+        } else {
+          GST_INFO_OBJECT(self, "set downscale width height :%d %d", self->downscale_width,self->downscale_height);
+        }
+      }
+    }
+#endif
+
   }
 #ifdef USE_OMX_TARGET_ZYNQ_USCALE_PLUS
   if (!set_zynqultrascaleplus_props (self))
@@ -1451,6 +1504,14 @@ gst_omx_video_enc_set_property (GObject * object, guint prop_id,
     case PROP_INTRA_REFRESH_MBS:
       self->intra_refresh_mbs = g_value_get_uint (value);
       break;
+#ifdef _USE_TARGET_VPU554_
+    case PROP_DOWNSCALE_WIDTH:
+      self->downscale_width= g_value_get_uint (value);
+      break;
+    case PROP_DOWNSCALE_HEIGHT:
+      self->downscale_height = g_value_get_uint (value);
+      break;
+#endif
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -1572,6 +1633,14 @@ gst_omx_video_enc_get_property (GObject * object, guint prop_id, GValue * value,
     case PROP_INTRA_REFRESH_MBS:
       g_value_set_uint (value, self->intra_refresh_mbs);
       break;
+#ifdef _USE_TARGET_VPU554_
+    case PROP_DOWNSCALE_WIDTH:
+      g_value_set_uint(value,self->downscale_width);
+      break;
+    case PROP_DOWNSCALE_HEIGHT:
+      g_value_set_uint(value,self->downscale_height);
+      break;
+#endif
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
diff --git a/omx/gstomxvideoenc.h b/omx/gstomxvideoenc.h
index 7f5f6da..d42e764 100644
--- a/omx/gstomxvideoenc.h
+++ b/omx/gstomxvideoenc.h
@@ -116,6 +116,10 @@ struct _GstOMXVideoEnc
   guint32 mirror;
   guint32 intra_refresh_mode;
   guint32 intra_refresh_mbs;
+#ifdef _USE_TARGET_VPU554_
+  guint32 downscale_width;
+  guint32 downscale_height;
+#endif
 
   guint32 default_target_bitrate;
 
-- 
2.7.4

